bin/utest/pgrp.c

```diff
@@ -0,0 +1,136 @@
+#include <assert.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/wait.h>
+
+#include "utest.h"
+
+int test_setpgid(void) {
+  /* Init process can create and enter new process group. */
+  assert(!setpgid(1, 0));
+
+  pid_t pid = fork();
+  if (pid == 0) {
+    /* Process inherits group of its parent. */
+    assert(getpgid(0) == 1);
+    /* Process can make its own group. */
+    assert(!setpgid(0, 0));
+    /* New group has ID equal to the ID of the moved process. */
+    assert(getpgid(0) == getpid());
+
+    exit(0);
+  }
+  wait(NULL);
+
+  /* It is forbidden to move the process to non-existing group. */
+  assert(setpgid(0, pid));
+  /* It is forbidden to change the group of non-existing process. */
+  assert(setpgid(pid, pid));
+
+  return 0;
+}
+
+static int sig_delivered = 0;
+
+static void sa_handler(int signo) {
+  assert(signo == SIGUSR1);
+  sig_delivered = 1;
+}
+
+static void kill_tests_setup(void) {
+  sig_delivered = 0;
+  setpgid(0, 1);
+
+  sigaction_t newact;
+  sigaction_t oldact;
+
+  newact.sa_handler = sa_handler;
+  newact.sa_restorer = sigreturn;
```

Comment:
Not that long ago we decided to remove `sa_restorer` from `struct sigaction` since it's not a part of POSIX.1 standard.